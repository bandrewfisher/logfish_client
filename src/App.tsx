import React, { useState, useEffect } from 'react';
import CopyToClipboard from 'react-copy-to-clipboard';

import {
  HashRouter as Router,
  Switch,
  Route,
} from 'react-router-dom';

import socket from './socket';
import NavBar from './components/NavBar';
import Button from './components/base/Button';
import Sidebar from './components/Sidebar';
import PhoneModal from './components/PhoneModal';
import TurnOffPhoneModal from './components/PhoneModal/TurnOffPhoneModal';
import Toast from './components/base/Toast';
import About from './components/About';

const App = () => {
  const [apiKey, setApiKey] = useState<string>('');
  const [logs, setLogs] = useState<string[]>([]);
  const [phoneModalOpen, setPhoneModalOpen] = useState<boolean>(false);
  const [turnOffPhoneModalOpen, setTurnOffPhoneModalOpen] = useState<boolean>(false);
  const [copyToastOpen, setCopyToastOpen] = useState<boolean>(false);
  const [isPhoneSetup, setIsPhoneSetup] = useState<boolean>(false);

  useEffect(() => {
    socket.on('CONNECT', (key: string) => {
      setApiKey(key);
    });
    socket.on('DATA', (data: any) => {
      setLogs((l) => [data, ...l]);
    });
  }, []);

  const copyApiKey = () => {
    setCopyToastOpen(true);
  };

  return (
    <Router>
      <div className="pt-24">
        <NavBar />
        <Switch>
          <Route exact path="/about">
            <About />
          </Route>
          <Route path="/">
            <div className="w-full flex">
              <Sidebar />
              <div className="flex-grow pr-6">
                <h2 className="text-2xl font-semibold text-gray-700 mb-3">
                  Your temporary API key (click to copy):
                </h2>
                <CopyToClipboard text={apiKey} onCopy={copyApiKey}>
                  <div className="rounded bg-green-200 p-3 border border-green-400 mb-8 cursor-pointer">
                    {apiKey}
                  </div>
                </CopyToClipboard>
                <div className="flex">
                  <div className="mr-4">
                    <h5 className="text-lg text-gray-700">
                      Your logs will appear below!
                    </h5>
                    <h5 className="text-lg text-gray-700">
                      Or try logging to your phone!
                    </h5>
                  </div>
                  {isPhoneSetup ? (
                    <Button
                      handleClick={() => setTurnOffPhoneModalOpen(true)}
                    >
                      Turn off phone logging
                    </Button>
                  ) : (
                    <Button handleClick={() => setPhoneModalOpen(true)}>
                      Set up phone logging
                    </Button>
                  )}
                </div>
                <textarea
                  className="w-full bg-gray-200 rounded h-full mt-6"
                  readOnly
                  value={logs.join('\n')}
                />
              </div>
              <PhoneModal
                open={phoneModalOpen}
                handleClose={() => setPhoneModalOpen(false)}
                onSetupSuccess={() => {
                  setIsPhoneSetup(true);
                }}
              />
              <TurnOffPhoneModal
                open={turnOffPhoneModalOpen}
                handleClose={() => setTurnOffPhoneModalOpen(false)}
                handleTurnOffPhone={() => setIsPhoneSetup(false)}
              />
            </div>
            <Toast open={copyToastOpen} text="Copied!" onClose={() => setCopyToastOpen(false)} />
          </Route>
        </Switch>
      </div>
    </Router>
  );
};
export default App;
