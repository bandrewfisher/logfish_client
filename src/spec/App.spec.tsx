import React from 'react';
import {
  render, RenderResult, wait, act,
} from '@testing-library/react';
import App from '../App';

let onCb: ((data: any) => void) | null = null;
function mockRegisterDataCb(cb: (data: any) => void) {
  onCb = cb;
}
let onConnect: ((data: any) => void) | null = null;
function mockRegisterConnectCb(cb: (data: any) => void) {
  onConnect = cb;
}
jest.mock('../socket', () => ({
  __esModule: true,
  default: {
    on: jest.fn((event: string, cb: (data: any) => void) => {
      if (event === 'CONNECT') {
        mockRegisterConnectCb(cb);
      } else if (event === 'DATA') {
        mockRegisterDataCb(cb);
      }
    }),
  },
}));

describe('App', () => {
  let wrapper: RenderResult;

  beforeEach(() => {
    wrapper = render(<App />);
  });

  test('matches snapshot', () => {
    const { container } = wrapper;
    expect(container).toMatchSnapshot();
  });
  test('displays API key', async () => {
    const { queryByText } = wrapper;
    act(() => {
      if (onConnect) {
        onConnect('FAKE_KEY');
      }
    });
    await wait(() => expect(queryByText('FAKE_KEY')).not.toBe(null));
  });
  test('registers socket connect', async () => {
    const { queryByText } = wrapper;
    act(() => {
      if (onCb) {
        onCb('MESSAGE');
      }
    });
    await wait(() => expect(queryByText('MESSAGE')).not.toBe(null));
  });
});
