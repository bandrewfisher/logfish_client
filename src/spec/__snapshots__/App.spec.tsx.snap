// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div>
  <div
    class="pt-24"
  >
    <div
      class="fixed top-0 inset-x-0 bg-transparent h-24 flex justify-between p-6 text-xl"
    >
      <div
        class="font-bold"
      >
        logfish
      </div>
      <div
        class="flex"
      >
        <div
          class="font-semibold text-gray-500 hover:text-blue-600 cursor-pointer mr-3"
        >
          about
        </div>
        <div
          class="font-semibold text-gray-500 hover:text-blue-600 cursor-pointer"
        >
          support
        </div>
      </div>
    </div>
    <div
      class="w-full flex"
    >
      <div
        class="px-6"
      >
        <h3
          class="text-gray-600 font-semibold text-xl mb-6"
        >
          How's it work?
        </h3>
        <ul>
          <li>
            1. Download the npm package.
          </li>
          <li>
            <div
              class="bg-gray-200 rounded-md border-gray-300 border p-2 inline-block"
            >
              <code>
                $ npm install logfish
              </code>
            </div>
          </li>
          <li>
            2. Create a logfish object.
          </li>
          <li>
            <div
              class="bg-gray-200 rounded-md border-gray-300 border p-2 inline-block"
            >
              <code>
                const lf = require('logfish').default;
                <br />
                // or 
                <br />
                import lf from 'logfish';
              </code>
            </div>
          </li>
          <li>
            3. Copy your API key.
          </li>
          <li>
            4. Log!
          </li>
        </ul>
      </div>
      <div
        class="flex-grow pr-6"
      >
        <h2
          class="text-2xl font-semibold text-gray-700 mb-3"
        >
          Your temporary API key (click to copy):
        </h2>
        <div
          class="rounded bg-green-200 p-3 border border-green-400 mb-8 cursor-pointer"
        />
        <div
          class="flex"
        >
          <div
            class="mr-4"
          >
            <h5
              class="text-lg text-gray-700"
            >
              Your logs will appear below!
            </h5>
            <h5
              class="text-lg text-gray-700"
            >
              Or try logging to your phone!
            </h5>
          </div>
          <button
            class="rounded-md p-3 text-white bg-blue-500"
            type="button"
          >
            Set up phone logging
          </button>
        </div>
        <textarea
          class="w-full bg-gray-200 rounded h-full mt-6"
          readonly=""
        />
      </div>
    </div>
    <div
      class="toast-container fixed flex justify-center bottom-0"
    >
      <div
        class="toast p-2 rounded-lg bg-gray-800 text-white toast-hide"
      >
        Copied!
      </div>
    </div>
  </div>
</div>
`;
